### PR TITLE
add mips r6 support

### PR DESCRIPTION
**Pull request by sebastianas**
mips r6 is quite similar with current mips architectures,
while mips r6 drops and replace some insn,
so they are not compatible with previous version.

Here for openssl, we disable ASM, and -DBN_DIV3W.

Debian BTS https://bugs.debian.org/882007
Signed-off-by: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>

- [X] documentation is added or updated
- [X] tests are added or updated

### CODE REVIEW

**Comment by sebastianas** [comment]
can this go via minor/obvious change so no CLA needed, please?

**Comment by dot-asm** [comment]
??? MIPS assembler pack was adapted for R6... Well, I didn't have opportunity to test Linux build, only on Android. Have you tried ./config -mips32r6 [or -mips64r6]?

**Comment by dot-asm** [comment]
Though R6 support is available only in master, while Debian report is about 1.1.0...

**Comment by dot-asm** [comment]
Provided that you don't pass -mips[32|64]r6 flags compiler must be assuming it by default. In such case what prevents you from simply using currently available targets with no-asm? I mean in 1.1.0, master should work as is.

**Comment by richsalz** [comment]
So adding a new platform to an existing release is adding a feature, not fixing a bug. And it's already in master, so close this?

**Comment by dot-asm** [comment]
To summarize. It should be possible to compile 1.1.0 as is with additional no-asm parameter, and suggestion is to do just that. [There are provisions that allow package maintainer to specify additional flags, aren't there?] As for master, it should be possible to compile as is, and it's not a coincidence, master aims to support R6.

On related note it should also be noted that target descriptions don't have to reside in 10-main.cf, one can as well put them into separate file. This is meant to be *preferred* way for site-specific adaptations that facilitates their cross-release maintenance. By last part I mean that maintaining separate add-on cf file is simpler and more reliable than tracking modifications in 10-main.cf file if site-specific adaption was maintained as private patch. Or in other words this is also something a package maintainer can do. Another problem that separate cf files intend to solve is community-maintained targets. In practical terms it means that if there is target that we have no way to judge, explicit or implicit, then we would prefer (insist really) to see it in separate cf file (if possible with with reference to maintainer).

Bottom line is that there are other preferred ways to solve the problem, so that yes, this is to be closed.

**Comment by sebastianas** [comment]
On 2017-11-29 10:49:18 [+0000], Andy Polyakov wrote:
> To summarize. It should be possible to compile 1.1.0 as is with
> additional no-asm parameter, and suggestion is to do just that. [There
> are provisions that allow package maintainer to specify additional
> flags, aren't there?] As for master, it should be possible to compile
> as is, and it's not a coincidence, master aims to support R6.

Ah okay. So on master you don't have a r6 target but aes-mips.pl
generates proper asm code. Good.

> On related note it should also be noted that target descriptions don't
> have to reside in 10-main.cf, one can as well put them into separate
> file. This is meant to be *preferred* way for site-specific
We do have that 20-debian.conf and do it that way but I *assumed* that
R6 will be mroe complicated than just what you did to make the asm
working :)

Sebastian

**Comment by dot-asm** [comment]
> So on master you don't have a r6 target but aes-mips.pl generates proper asm code.

Just in case, to eliminate ambiguity. MIPS assembly modules emit same code in all cases. What happens is that output is then pre-processed by C pre-processor, and I rely on C compiler driver to pre-define processor-specific macros. In other words it's not like [master] aes-mips.pl generates proper R6 code, it generates all possible variants, and suitable one is chosen by C pre-processor depending on compiler arch selection flag or default arch setting.
